### PR TITLE
oomd: fix memory leak

### LINKED ISSUE #18926: Possible memory leak in systemd-oomd
**systemd version the issue has been seen with**
 > latest master

**Used distribution**
 > Fedora 33

Reproducible by running `TEST-55-OOMD` under sanitizers:

```
# meson build -Dtests=true -Dinstall-tests=true -Db_sanitize=address,undefined
# meson compile -C build
# make -C test/TEST-55-OOMD/ clean setup run SKIP_INITRD=yes
```

```
[  147.736800] systemd-oomd[310]: =================================================================
[  147.737197] systemd-oomd[310]: ==310==ERROR: LeakSanitizer: detected memory leaks
[  147.737663] systemd-oomd[310]: Direct leak of 120 byte(s) in 1 object(s) allocated from:
[  147.738085] systemd-oomd[310]:     #0 0x7fee4bad1587 in __interceptor_calloc (/usr/lib64/libasan.so.6.0.0+0xab587)
[  147.738626] systemd-oomd[310]:     #1 0x40fbc8 in oomd_cgroup_context_acquire ../src/oom/oomd-util.c:271
[  147.739045] systemd-oomd[310]:     #2 0x41173d in oomd_insert_cgroup_context ../src/oom/oomd-util.c:393
[  147.739611] systemd-oomd[310]:     #3 0x40639f in process_managed_oom_reply ../src/oom/oomd-manager.c:115
[  147.740027] systemd-oomd[310]:     #4 0x7fee4a9980d4 in varlink_dispatch_reply ../src/shared/varlink.c:749
[  147.740480] systemd-oomd[310]:     #5 0x7fee4a999989 in varlink_process ../src/shared/varlink.c:940
[  147.740933] systemd-oomd[310]:     #6 0x7fee4a9a540c in defer_callback ../src/shared/varlink.c:1862
[  147.741386] systemd-oomd[310]:     #7 0x7fee4ae00a19 in source_dispatch ../src/libsystemd/sd-event/sd-event.c:3523
[  147.744885] systemd-oomd[310]:     #8 0x7fee4ae0792b in sd_event_dispatch ../src/libsystemd/sd-event/sd-event.c:4004
[  147.745265] systemd-oomd[310]:     #9 0x7fee4ae089fd in sd_event_run ../src/libsystemd/sd-event/sd-event.c:4065
[  147.745757] systemd-oomd[310]:     #10 0x7fee4ae08d49 in sd_event_loop ../src/libsystemd/sd-event/sd-event.c:4086
[  147.751448] systemd-oomd[310]:     #11 0x417aba in run ../src/oom/oomd.c:175
[  147.751717] systemd-oomd[310]:     #12 0x417c60 in main ../src/oom/oomd.c:182
[  147.751894] systemd-oomd[310]:     #13 0x7fee493721e1 in __libc_start_main (/lib64/libc.so.6+0x281e1)
[  147.752113] systemd-oomd[310]: Indirect leak of 64 byte(s) in 1 object(s) allocated from:
[  147.752340] systemd-oomd[310]:     #0 0x7fee4ba7e547 in strdup (/usr/lib64/libasan.so.6.0.0+0x58547)
[  147.752560] systemd-oomd[310]:     #1 0x410d05 in oomd_cgroup_context_acquire ../src/oom/oomd-util.c:333
[  147.752872] systemd-oomd[310]:     #2 0x41173d in oomd_insert_cgroup_context ../src/oom/oomd-util.c:393
[  147.753101] systemd-oomd[310]:     #3 0x40639f in process_managed_oom_reply ../src/oom/oomd-manager.c:115
[  147.753315] systemd-oomd[310]:     #4 0x7fee4a9980d4 in varlink_dispatch_reply ../src/shared/varlink.c:749
[  147.753529] systemd-oomd[310]:     #5 0x7fee4a999989 in varlink_process ../src/shared/varlink.c:940
[  147.753744] systemd-oomd[310]:     #6 0x7fee4a9a540c in defer_callback ../src/shared/varlink.c:1862
[  147.753960] systemd-oomd[310]:     #7 0x7fee4ae00a19 in source_dispatch ../src/libsystemd/sd-event/sd-event.c:3523
[  147.754177] systemd-oomd[310]:     #8 0x7fee4ae0792b in sd_event_dispatch ../src/libsystemd/sd-event/sd-event.c:4004
[  147.754415] systemd-oomd[310]:     #9 0x7fee4ae089fd in sd_event_run ../src/libsystemd/sd-event/sd-event.c:4065
[  147.754624] systemd-oomd[310]:     #10 0x7fee4ae08d49 in sd_event_loop ../src/libsystemd/sd-event/sd-event.c:4086
[  147.754864] systemd-oomd[310]:     #11 0x417aba in run ../src/oom/oomd.c:175
[  147.755080] systemd-oomd[310]:     #12 0x417c60 in main ../src/oom/oomd.c:182
[  147.755304] systemd-oomd[310]:     #13 0x7fee493721e1 in __libc_start_main (/lib64/libc.so.6+0x281e1)
[  147.755521] systemd-oomd[310]: SUMMARY: AddressSanitizer: 184 byte(s) leaked in 2 allocation(s).
```

/cc @anitazha

### PR DESCRIPTION
**Pull request by anitazha**
Fixes #18926 

cc @mrc0mmand

### CODE REVIEW

File: src/oom/oomd-util.c [hunk 1]
```diff
@@ -385,24 +385,29 @@ int oomd_system_context_acquire(const char *proc_swaps_path, OomdSystemContext *
...
 
-        old_ctx = hashmap_get(old_h, path);
+        old_ctx = hashmap_get(old_h, norm_path);
         if (old_ctx) {
                 curr_ctx->last_pgscan = old_ctx->pgscan;
                 curr_ctx->mem_pressure_limit = old_ctx->mem_pressure_limit;
                 curr_ctx->last_hit_mem_pressure_limit = old_ctx->last_hit_mem_pressure_limit;
         }
 
+        assert_se(streq(norm_path, curr_ctx->path));
```
Comment (keszybz) [inline src/oom/oomd-util.c]: `assert`, without `_se`.

**Comment by mrc0mmand** [comment]
Thanks! As for the Packit fails - if you rebase to the latest ~~master~~ main, they should go away.

**Review by keszybz — COMMENTED** [review]
LGTM except for that one thing.

**Comment by anitazha** [comment]
Thanks folks. Rebased and changed the one assert_se to assert

**Comment by mrc0mmand** [comment]
The current CentOS CI fail is most likely caused by this:

```
[    3.965247] testsuite-55.sh[321]: + oomctl
Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/systemd_2dupdate_2dutmp_2drunlevel_2eservice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=135 reply_cookie=0 signature=saa
[    3.967114] testsuite-55.sh[322]: + grep /testsuite-55-workload.slice
Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/systemd_2dupdate_2dutmp_2drunlevel_2eservice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=136 reply_cookie=0 signature=saa
Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=JobRemoved cookie=137 reply_cookie=0 signature=uoss error-name=n/a error-message=n/a
systemd-update-utmp-runlevel.service: Consumed 51ms CPU time.
Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/systemd_2dupdate_2dutmp_2drunlevel_2eservice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=138 reply_cookie=0 signature=saa
Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/systemd_2dupdate_2dutmp_2drunlevel_2eservice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=139 reply_cookie=0 signature=saa
systemd-update-utmp-runlevel.service: Control group is empty.
Bus private-bus-connection: changing state CLOSING → CLOSED
Got disconnect on private connection.
systemd-journald.service: Received EPOLLHUP on stored fd 63 (stored), closing.
[    3.988008] systemd-oomd[300]: =================================================================
[    3.988752] systemd-oomd[300]: ==300==ERROR: AddressSanitizer: heap-use-after-free on address 0x606000001d00 at pc 0x7fd740f4059c bp 0x7fffb53b8840 sp 0x7fffb53b7ff0
[    3.989204] systemd-oomd[300]: READ of size 2 at 0x606000001d00 thread T0
[    4.031115] systemd-oomd[300]:     #0 0x7fd740f4059b  (/usr/lib64/libasan.so.6.0.0+0x8a59b)
[    4.033470] systemd-oomd[300]:     #1 0x7fd73fef269a in base_bucket_scan ../src/basic/hashmap.c:1219
[    4.033933] systemd-oomd[300]:     #2 0x7fd73fef369e in _hashmap_remove ../src/basic/hashmap.c:1397
[    4.034150] systemd-oomd[300]:     #3 0x4052d8 in hashmap_remove ../src/basic/hashmap.h:196
[    4.034401] systemd-oomd[300]:     #4 0x40616e in process_managed_oom_reply ../src/oom/oomd-manager.c:98
[    4.036389] systemd-oomd[300]:     #5 0x7fd73fe280d4 in varlink_dispatch_reply ../src/shared/varlink.c:749
[    4.036899] systemd-oomd[300]:     #6 0x7fd73fe29989 in varlink_process ../src/shared/varlink.c:940
[    4.037268] systemd-oomd[300]:     #7 0x7fd73fe3540c in defer_callback ../src/shared/varlink.c:1862
[    4.038266] systemd-oomd[300]:     #8 0x7fd740290a59 in source_dispatch ../src/libsystemd/sd-event/sd-event.c:3523
[    4.038602] systemd-oomd[300]:     #9 0x7fd74029796b in sd_event_dispatch ../src/libsystemd/sd-event/sd-event.c:4004
[    4.038954] systemd-oomd[300]:     #10 0x7fd740298a3d in sd_event_run ../src/libsystemd/sd-event/sd-event.c:4065
[    4.039352] systemd-oomd[300]:     #11 0x7fd740298d89 in sd_event_loop ../src/libsystemd/sd-event/sd-event.c:4086
[    4.039557] systemd-oomd[300]:     #12 0x417b49 in run ../src/oom/oomd.c:175
[    4.039789] systemd-oomd[300]:     #13 0x417cef in main ../src/oom/oomd.c:182
[    4.040025] systemd-oomd[300]:     #14 0x7fd73e8021e1 in __libc_start_main (/lib64/libc.so.6+0x281e1)
[    4.040419] systemd-oomd[300]:     #15 0x404add in _start (/usr/lib/systemd/systemd-oomd+0x404add)
[    4.040756] systemd-oomd[300]: 0x606000001d00 is located 0 bytes inside of 64-byte region [0x606000001d00,0x606000001d40)
[    4.041116] systemd-oomd[300]: freed by thread T0 here:
[    4.041363] systemd-oomd[300]:     #0 0x7fd740f610c7 in __interceptor_free (/usr/lib64/libasan.so.6.0.0+0xab0c7)
[    4.041710] systemd-oomd[300]:     #1 0x405a25 in managed_oom_reply_destroy ../src/oom/oomd-manager.c:24
[    4.042027] systemd-oomd[300]:     #2 0x4064d3 in process_managed_oom_reply ../src/oom/oomd-manager.c:77
[    4.042390] systemd-oomd[300]:     #3 0x7fd73fe280d4 in varlink_dispatch_reply ../src/shared/varlink.c:749
[    4.042741] systemd-oomd[300]:     #4 0x7fd73fe29989 in varlink_process ../src/shared/varlink.c:940
[    4.042984] systemd-oomd[300]:     #5 0x7fd73fe3540c in defer_callback ../src/shared/varlink.c:1862
[    4.043233] systemd-oomd[300]:     #6 0x7fd740290a59 in source_dispatch ../src/libsystemd/sd-event/sd-event.c:3523
[    4.043581] systemd-oomd[300]:     #7 0x7fd74029796b in sd_event_dispatch ../src/libsystemd/sd-event/sd-event.c:4004
[    4.043920] systemd-oomd[300]:     #8 0x7fd740298a3d in sd_event_run ../src/libsystemd/sd-event/sd-event.c:4065
[    4.044263] systemd-oomd[300]:     #9 0x7fd740298d89 in sd_event_loop ../src/libsystemd/sd-event/sd-event.c:4086
[    4.044614] systemd-oomd[300]:     #10 0x417b49 in run ../src/oom/oomd.c:175
[    4.044852] systemd-oomd[300]:     #11 0x417cef in main ../src/oom/oomd.c:182
[    4.045088] systemd-oomd[300]:     #12 0x7fd73e8021e1 in __libc_start_main (/lib64/libc.so.6+0x281e1)
[    4.045449] systemd-oomd[300]: previously allocated by thread T0 here:
[    4.045661] systemd-oomd[300]:     #0 0x7fd740f0e547 in strdup (/usr/lib64/libasan.so.6.0.0+0x58547)
[    4.046026] systemd-oomd[300]:     #1 0x7fd73ff96a64 in free_and_strdup ../src/basic/string-util.c:867
[    4.046243] systemd-oomd[300]:     #2 0x7fd73fd48dce in json_dispatch_string ../src/shared/json.c:4142
[    4.046589] systemd-oomd[300]:     #3 0x7fd73fd478b5 in json_dispatch ../src/shared/json.c:3985
[    4.046913] systemd-oomd[300]:     #4 0x405f5d in process_managed_oom_reply ../src/oom/oomd-manager.c:86
[    4.047253] systemd-oomd[300]:     #5 0x7fd73fe280d4 in varlink_dispatch_reply ../src/shared/varlink.c:749
[    4.047605] systemd-oomd[300]:     #6 0x7fd73fe29989 in varlink_process ../src/shared/varlink.c:940
[    4.047939] systemd-oomd[300]:     #7 0x7fd73fe3540c in defer_callback ../src/shared/varlink.c:1862
[    4.048176] systemd-oomd[300]:     #8 0x7fd740290a59 in source_dispatch ../src/libsystemd/sd-event/sd-event.c:3523
[    4.048418] systemd-oomd[300]:     #9 0x7fd74029796b in sd_event_dispatch ../src/libsystemd/sd-event/sd-event.c:4004
[    4.048758] systemd-oomd[300]:     #10 0x7fd740298a3d in sd_event_run ../src/libsystemd/sd-event/sd-event.c:4065
[    4.049112] systemd-oomd[300]:     #11 0x7fd740298d89 in sd_event_loop ../src/libsystemd/sd-event/sd-event.c:4086
[    4.049465] systemd-oomd[300]:     #12 0x417b49 in run ../src/oom/oomd.c:175
[    4.049794] systemd-oomd[300]:     #13 0x417cef in main ../src/oom/oomd.c:182
[    4.050034] systemd-oomd[300]:     #14 0x7fd73e8021e1 in __libc_start_main (/lib64/libc.so.6+0x281e1)
[    4.050262] systemd-oomd[300]: SUMMARY: AddressSanitizer: heap-use-after-free (/usr/lib64/libasan.so.6.0.0+0x8a59b)
[    4.050616] systemd-oomd[300]: Shadow bytes around the buggy address:
[    4.050954] systemd-oomd[300]:   0x0c0c7fff8350: fd fd fd fd fa fa fa fa fd fd fd fd fd fd fd fd
[    4.051177] systemd-oomd[300]:   0x0c0c7fff8360: fa fa fa fa fd fd fd fd fd fd fd fd fa fa fa fa
[    4.051417] systemd-oomd[300]:   0x0c0c7fff8370: fd fd fd fd fd fd fd fd fa fa fa fa fd fd fd fd
[    4.051751] systemd-oomd[300]:   0x0c0c7fff8380: fd fd fd fd fa fa fa fa fd fd fd fd fd fd fd fd
[    4.051984] systemd-oomd[300]:   0x0c0c7fff8390: fa fa fa fa fd fd fd fd fd fd fd fd fa fa fa fa
[    4.052221] systemd-oomd[300]: =>0x0c0c7fff83a0:[fd]fd fd fd fd fd fd fd fa fa fa fa fd fd fd fd
[    4.052463] systemd-oomd[300]:   0x0c0c7fff83b0: fd fd fd fd fa fa fa fa fd fd fd fd fd fd fd fd
[    4.052787] systemd-oomd[300]:   0x0c0c7fff83c0: fa fa fa fa fd fd fd fd fd fd fd fd fa fa fa fa
[    4.053020] systemd-oomd[300]:   0x0c0c7fff83d0: fd fd fd fd fd fd fd fd fa fa fa fa fd fd fd fd
[    4.053250] systemd-oomd[300]:   0x0c0c7fff83e0: fd fd fd fd fa fa fa fa fd fd fd fd fd fd fd fd
[    4.053600] systemd-oomd[300]:   0x0c0c7fff83f0: fa fa fa fa fd fd fd fd fd fd fd fd fa fa fa fa
[    4.053930] systemd-oomd[300]: Shadow byte legend (one shadow byte represents 8 application bytes):
[    4.054155] systemd-oomd[300]:   Addressable:           00
[    4.054400] systemd-oomd[300]:   Partially addressable: 01 02 03 04 05 06 07
[    4.054726] systemd-oomd[300]:   Heap left redzone:       fa
[    4.054971] systemd-oomd[300]:   Freed heap region:       fd
[    4.055186] systemd-oomd[300]:   Stack left redzone:      f1
[    4.055425] systemd-oomd[300]:   Stack mid redzone:       f2
[    4.055754] systemd-oomd[300]:   Stack right redzone:     f3
[    4.055990] systemd-oomd[300]:   Stack after return:      f5
[    4.056331] systemd-oomd[300]:   Stack use after scope:   f8
[    4.056566] systemd-oomd[300]:   Global redzone:          f9
[    4.056892] systemd-oomd[300]:   Global init order:       f6
[    4.057126] systemd-oomd[300]:   Poisoned by user:        f7
[    4.057363] systemd-oomd[300]:   Container overflow:      fc
[    4.057591] systemd-oomd[300]:   Array cookie:            ac
[    4.057927] systemd-oomd[300]:   Intra object redzone:    bb
[    4.058153] systemd-oomd[300]:   ASan internal:           fe
[    4.058408] systemd-oomd[300]:   Left alloca redzone:     ca
[    4.058734] systemd-oomd[300]:   Right alloca redzone:    cb
[    4.058973] systemd-oomd[300]:   Shadow gap:              cc
[    4.059200] systemd-oomd[300]: ==300==ABORTING
```

**Comment by anitazha** [comment]
Ugh I know what this is.. fix coming

File: src/oom/oomd-util.c [hunk 2]
```diff
@@ -385,16 +385,21 @@ int oomd_system_context_acquire(const char *proc_swaps_path, OomdSystemContext *
...
         _cleanup_(oomd_cgroup_context_freep) OomdCGroupContext *curr_ctx = NULL;
         OomdCGroupContext *old_ctx;
+        const char *norm_path;
         int r;
 
         assert(new_h);
         assert(path);
 
-        r = oomd_cgroup_context_acquire(path, &curr_ctx);
+        norm_path = empty_to_root(path);
```
Comment (yuwata) [inline src/oom/oomd-util.c]: Not a matter, but simply, `path = empty_to_root(path)` should work.

File: src/oom/oomd-util.c [hunk 3]
```diff
@@ -385,16 +385,21 @@ int oomd_system_context_acquire(const char *proc_swaps_path, OomdSystemContext *
...
+
+        r = oomd_cgroup_context_acquire(norm_path, &curr_ctx);
         if (r < 0)
-                return log_debug_errno(r, "Failed to get OomdCGroupContext for %s: %m", path);
+                return log_debug_errno(r, "Failed to get OomdCGroupContext for %s: %m", norm_path);
+
+        assert(streq(norm_path, curr_ctx->path));
```
Comment (keszybz) [inline src/oom/oomd-util.c]: This got reverted to the old version again.

I'll force-push with this change and @yuwata's suggestion.

**Comment by mrc0mmand** [comment]
CentOS CI seems to be happy and I can't trigger any ASan error locally (so far). Thanks!

**Pull request merged**